### ISSUE TITLE
Jax dependency

### ISSUE THREAD

**Issue by scarrazza** [qiboteam/qiboml#83]
From the project toml jax and jaxlib are installed by default.
Is it possible to convert both deps to optional?

**Comment by scarrazza** [comment]
Does the code works with just numpy? I considering users that are interested in building simple models, e.g. a VQE, without having to download and deal with AI packages.

**Comment by MatteoRobbiati** [comment]
> Does the code works with just numpy? I considering users that are interested in building simple models, e.g. a VQE, without having to download and deal with AI packages.

Right now we only support gradient-based optimization. To do so, the only option working with Numpy is the case of parameter shift rules. I wouldn't set it as a default because of computational cost, but for sure it is a possibility. 
Otherwise, we can setup a tutorial or take into account the possibility of integrating other optimizers.

**Comment by scarrazza** [comment]
We should discuss about the defaults, @renatomello might have some opinion about this too.

**Comment by renatomello** [comment]
So what I proposed to @scarrazza is that it would be better if somehow it was possible for `qiboml` to not install any of the three current engines by default. For instance, a lot of people @ TII are interested in using `qibo` + `torch` differentiation. In that case, it doesn't make much sense to forcibly install `jax`. One can easily think of other scenarios where someone is interested in one of the three engines but not the other two. I don't how would work in terms of the code, but we could think of a way (maybe implicit dependencies instead of explicit).

**Comment by BrunoLiegiBastonLiegi** [comment]
Well, the fact is that ``qibo`` by default ships with ``numpy`` and its ``NumpyBackend``, and at the moment we only provide two methods to compute gradients outside of the native ``pytorch`` and ``tensorflow`` autodifferentiation engines: 
- ``qiboml.differentiation.Jax``, explicitely thought to be used with ``numpy``-based backends
- ``qiboml.differentiation.PSR``, mostly useful for hardware backends, e.g. ``qibolab``

If you remove the ``jax`` dependency, you are left with just the ``PSR`` which is way slower and not very practical. Therefore, in order to have ``qiboml`` fully functional with a standard ``qibo`` installation it would be better to have ``jax`` available, and I personally don't see major problems in having it installed alongside ``qiboml``.

**Comment by alecandido** [comment]
I agree with @scarrazza and @renatomello. It doesn't matter if it's practical or not, having JAX around should be an option. And since Python's infrastructure is lacking opt-out dependencies, it has to be opt-in.

> and I personally don't see major problems in having it installed alongside `qiboml`.

It's not much a matter of personal taste: any additional mandatory dependency makes it harder to manage the environment, running at risk to rule out some users.
This is acceptable if there is no way to work without it. But it is an arbitrary additional limitation if not.

If you can use the PSR with nothing else other than Qibo and NumPy, then there should be no other mandatory dependencies.

If at any time someone finds too complex to use `pip install qiboml[jax]`, we can solve that issue in other ways. E.g. publishing a `qiboml-jax` package, which would turn the former into `pip install qiboml-jax`.
But UI issues should not interfere with the maintenance of functional features (both for code and dependencies).

**Comment by renatomello** [comment]
> I agree with [@scarrazza](https://github.com/scarrazza) and [@renatomello](https://github.com/renatomello). It doesn't matter if it's practical or not, having JAX around should be an option. And since Python's infrastructure is lacking opt-out dependencies, it has to be opt-in.
> 
> > and I personally don't see major problems in having it installed alongside `qiboml`.
> 
> It's not much a matter of personal taste: any additional mandatory dependency makes it harder to manage the environment, running at risk to rule out some users. This is acceptable if there is no way to work without it. But it is an arbitrary additional limitation if not.

Thinking about it, the same issue appears in `qibojit` as well since `numba` is installed by default even if one only plans on using GPUs.
